source/opt/decoration_manager.cpp

```diff
@@ -231,31 +231,78 @@ std::vector<T> DecorationManager::InternalGetDecorationsFor(
 
 void DecorationManager::ForEachDecoration(uint32_t id,
                                           uint32_t decoration,
-                                          std::function<void(const ir::Instruction&)> f) const {
-  auto decoration_list = id_to_decoration_insts_.find(id);
-  if (decoration_list != id_to_decoration_insts_.end()) {
-    for (const ir::Instruction* inst : decoration_list->second) {
-      switch (inst->opcode()) {
-        case SpvOpDecorate:
-          if (inst->GetSingleWordInOperand(1) == decoration) {
-            f(*inst);
-          }
-          break;
-        case SpvOpMemberDecorate:
-          if (inst->GetSingleWordInOperand(2) == decoration) {
-            f(*inst);
-          }
-          break;
-        case SpvOpDecorateId:
-          if (inst->GetSingleWordInOperand(1) == decoration) {
-            f(*inst);
+                                          std::function<void(const ir::Instruction&)> f) {
+  for (const ir::Instruction* inst : GetDecorationsFor(id, false)) {
+    switch (inst->opcode()) {
+      case SpvOpDecorate:
+        if (inst->GetSingleWordInOperand(1) == decoration) {
+          f(*inst);
+        }
+        break;
+      case SpvOpMemberDecorate:
+        if (inst->GetSingleWordInOperand(2) == decoration) {
+          f(*inst);
+        }
+        break;
+      case SpvOpDecorateId:
+        if (inst->GetSingleWordInOperand(1) == decoration) {
+          f(*inst);
+        }
+        break;
+      default:
+        assert(false && "Unexpected decoration instruction");
+    }
+  }
+}
+
+void DecorationManager::CloneDecorations(uint32_t from, uint32_t to) {
+  auto const decoration_list = id_to_decoration_insts_.find(from);
+  if (decoration_list == id_to_decoration_insts_.end()) return;
+  for (ir::Instruction* inst : decoration_list->second) {
+    switch (inst->opcode()) {
+      case SpvOpGroupDecorate:
+        // add |to| to list of decorated id's
+        inst->AddOperand(ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID, {to}));
+        id_to_decoration_insts_[to].push_back(inst);
+        break;
+      case SpvOpGroupMemberDecorate: {
+        bool is_id = false; // we start with decoration group id, followed by target-id, literal, target-id...
+        bool to_add = false;
+        // for each (id == from), get literal and add (to, literal) as operands
+        inst->ForEachInOperand([&](uint32_t* operand) {
```

Comment:
I don't know what other people think, but my preference is to have a loop where 1 iteration of the loop treat the pair instead of using a boolean that keeps flipping and the use of to_add.  I would suggest:
```
for( int op = <first operand>; op < inst.NumOperands(); op += 2 ) {
  if ( <id is "from"> ) {
    // Add pair of operands <to, next literal>.
  }
}
```
To me this is easier to read.  @dneto0, do you have a preference for this type of code?

Reply:
I am fine with both but would have preferred paired operands each of two words ;)